File: src/main/java/com/machiav3lli/backup/handler/SortFilterManager.java
package com.machiav3lli.backup.handler;

import android.content.Context;

import com.machiav3lli.backup.Constants;
import com.machiav3lli.backup.items.SortFilterModel;
import com.machiav3lli.backup.items.AppInfo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class SortFilterManager {

    public static SortFilterModel getFilterPreferences(Context context) {
        SortFilterModel sortFilterModel;
        if (!Utils.getPrefsString(context, Constants.PREFS_SORT_FILTER).equals(""))
            sortFilterModel = new SortFilterModel(Utils.getPrefsString(context, Constants.PREFS_SORT_FILTER));
        else sortFilterModel = new SortFilterModel();
        return sortFilterModel;
    }

    public static void saveFilterPreferences(Context context, SortFilterModel filterModel) {
        Utils.setPrefsString(context, Constants.PREFS_SORT_FILTER, filterModel.toString());
    }

    public static ArrayList<AppInfo> applyFilter(ArrayList<AppInfo> list, CharSequence filter, Context context) {
        ArrayList<AppInfo> nlist = (ArrayList<AppInfo>) list.clone();
        switch (filter.charAt(1)) {
            case '1':
                for (AppInfo item : list) if (!item.isSystem()) nlist.remove(item);
                break;
            case '2':
                for (AppInfo item : list) if (item.isSystem()) nlist.remove(item);
                break;
            default:
                break;
        }
        return applyOtherFilter(nlist, filter, context);
    }

    private static ArrayList<AppInfo> applyOtherFilter(ArrayList<AppInfo> list, CharSequence filter, Context context) {
        ArrayList<AppInfo> nlist = (ArrayList<AppInfo>) list.clone();
        switch (filter.charAt(2)) {
            case '1':
                for (AppInfo item : list) {
                    if (!(item.getLogInfo() == null ||
                            (item.getLogInfo().getVersionCode() != 0 &&
                                    item.getVersionCode() > item.getLogInfo().getVersionCode())))
                        nlist.remove(item);
                }
                break;
            case '2':
                for (AppInfo item : list) if (item.isInstalled()) nlist.remove(item);
                break;
            case '3':
                for (AppInfo item : list)
                    if (item.getLogInfo() != null) nlist.remove(item);
                break;
            case '4':
                int days;
                if (Utils.getPrefsString(context, Constants.PREFS_OLDBACKUPS).equals("")) days = 3;
                else days = Integer.parseInt(Utils.getPrefsString(context, Constants.PREFS_OLDBACKUPS));
                for (AppInfo item : list) {
                    if (item.getLogInfo() != null) {
                        long lastBackup = item.getLogInfo().getLastBackupMillis();
                        long diff = System.currentTimeMillis() - lastBackup;
                        if (!(lastBackup > 0 && diff > (days * 24 * 60 * 60 * 1000f)))
                            nlist.remove(item);
                    } else nlist.remove(item);
                }
                break;
            case '5':
                for (AppInfo item : list)
                    if (item.getBackupMode() != AppInfo.MODE_APK) nlist.remove(item);
                break;
            case '6':
                for (AppInfo item : list)
                    if (item.getBackupMode() != AppInfo.MODE_DATA) nlist.remove(item);
                break;
            case '7':
                for (AppInfo item : list) if (!item.isSpecial()) nlist.remove(item);
                break;
            default:
                break;
        }
        return applySort(nlist, filter);
    }

    private static ArrayList<AppInfo> applySort(ArrayList<AppInfo> list, CharSequence filter) {
        if (filter.charAt(0) == '0') Collections.sort(list, appInfoPackageNameComparator);
        else Collections.sort(list, appInfoLabelComparator);
        return list;
    }

    private static Comparator<AppInfo> appInfoLabelComparator = (m1, m2) ->
            m1.getLabel().compareToIgnoreCase(m2.getLabel());
    private static Comparator<AppInfo> appInfoPackageNameComparator = (m1, m2) ->
            m1.getPackageName().compareToIgnoreCase(m2.getPackageName());
}

File: app/src/main/java/com/machiav3lli/backup/handler/SortFilterManager.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.handler

import android.content.Context
import android.content.Intent
import com.machiav3lli.backup.*
import com.machiav3lli.backup.items.AppInfo
import com.machiav3lli.backup.utils.getDefaultSharedPreferences
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

object SortFilterManager {
    private val APP_INFO_LABEL_COMPARATOR = { m1: AppInfo, m2: AppInfo -> m1.packageLabel.compareTo(m2.packageLabel, ignoreCase = true) }
    private val APP_INFO_PACKAGE_NAME_COMPARATOR = { m1: AppInfo, m2: AppInfo -> m1.packageName.compareTo(m2.packageName, ignoreCase = true) }
    private val APP_INFO_DATA_SIZE_COMPARATOR = { m1: AppInfo, m2: AppInfo -> m1.dataBytes.compareTo(m2.dataBytes) }

    fun applyFilter(list: List<AppInfo>, filter: CharSequence, context: Context): List<AppInfo> {
        val predicate: (AppInfo) -> Boolean
        var launchableAppsList = listOf<String>()
        if (filter[1] == MAIN_FILTER_LAUNCHABLE) {
            val mainIntent = Intent(Intent.ACTION_MAIN, null).addCategory(Intent.CATEGORY_LAUNCHER)
            launchableAppsList = context.packageManager.queryIntentActivities(mainIntent, 0)
                    .map { it.activityInfo.packageName }
        }
        predicate = when (filter[1]) {
            MAIN_FILTER_SYSTEM -> { appInfo: AppInfo -> appInfo.isSystem }
            MAIN_FILTER_USER -> { appInfo: AppInfo -> !appInfo.isSystem }
            MAIN_FILTER_SPECIAL -> { appInfo: AppInfo -> appInfo.isSpecial }
            MAIN_FILTER_LAUNCHABLE -> { appInfo: AppInfo -> launchableAppsList.contains(appInfo.packageName) }
            else -> { _: AppInfo -> true }
        }
        val filteredList = list
                .filter(predicate)
                .toList()
        return applyBackupFilter(filteredList, filter, context)
    }

    private fun applyBackupFilter(list: List<AppInfo>, filter: CharSequence, context: Context): List<AppInfo> {
        val predicate: (AppInfo) -> Boolean = when (filter[2]) {
            MAIN_BACKUPFILTER_BOTH -> { appInfo: AppInfo -> appInfo.hasApk && appInfo.hasAppData }
            MAIN_BACKUPFILTER_APK -> { appInfo: AppInfo -> appInfo.hasApk }
            MAIN_BACKUPFILTER_DATA -> { appInfo: AppInfo -> appInfo.hasAppData }
            MAIN_BACKUPFILTER_NONE -> { appInfo: AppInfo -> !appInfo.hasBackups }
            else -> { _: AppInfo -> true }
        }
        val filteredList = list
                .filter(predicate)
                .toList()
        return applySpecialFilter(filteredList, filter, context)
    }

    private fun applySpecialFilter(list: List<AppInfo>, filter: CharSequence, context: Context): List<AppInfo> {
        val predicate: (AppInfo) -> Boolean
        val days = getDefaultSharedPreferences(context).getString(PREFS_OLDBACKUPS, "7")?.toInt()
                ?: 7
        predicate = when (filter[3]) {
            MAIN_SPECIALFILTER_NEW_UPDATED -> { appInfo: AppInfo -> !appInfo.hasBackups || appInfo.isUpdated }
            MAIN_SPECIALFILTER_NOTINSTALLED -> { appInfo: AppInfo -> !appInfo.isInstalled }
            MAIN_SPECIALFILTER_OLD -> {
                { appInfo: AppInfo ->
                    when {
                        appInfo.hasBackups -> {
                            val lastBackup = appInfo.latestBackup?.backupProperties?.backupDate
                            val diff = ChronoUnit.DAYS.between(lastBackup, LocalDateTime.now())
                            diff >= days
                        }
                        else -> false
                    }
                }
            }
            MAIN_SPECIALFILTER_SPLIT -> { appInfo: AppInfo -> appInfo.apkSplits.isNotEmpty() }
            else -> { _: AppInfo -> true }
        }
        val filteredList = list
                .filter(predicate)
                .toList()
        return applySort(filteredList, filter)
    }

    private fun applySort(list: List<AppInfo>, filter: CharSequence): List<AppInfo> {
        return when (filter[0]) {
            MAIN_SORT_PACKAGENAME -> list.sortedWith(APP_INFO_PACKAGE_NAME_COMPARATOR)
            MAIN_SORT_DATASIZE -> list.sortedWith(APP_INFO_DATA_SIZE_COMPARATOR)
            else -> list.sortedWith(APP_INFO_LABEL_COMPARATOR)
        }
    }
}